File: c++/59_Spiral_Matrix_II.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>
#include <cmath>

using namespace std;


class Solution {
public:
    vector<vector<int> > generateMatrix(int n) {
    	vector<vector<int> > res(n, vector<int>(n)); // initialize 2-d vector

    	int left = 0, top = 0;
    	int right = n - 1, down = n - 1;
    	int direction = 0;
    	int number = 1;
    	while (1) {
    		if (direction == 0) {
    			for (int i = left; i <= right; i++)
    				res[top][i] = number++;
    			top++;
    		}
    		if (direction == 1) {
    			for (int i = top; i <= down; i++)
    				res[i][right] = number++;
    			right--;
    		}
    		if (direction == 2) {
    			for (int i = right; i >= left; i--)
    				res[down][i] = number++;
    			down--;
    		}
    		if (direction == 3) {
    			for (int i = down; i >= top; i--)
    				res[i][left] = number++;
    			left++;
    		}
    		if (number == pow(n, 2) + 1)
    			break;
    		direction = (direction + 1) % 4;
    	}
    	return res;
    }
};


int main() {
	Solution s;
	vector<vector<int> > res = s.generateMatrix(3);

}

File: python/59_Spiral_Matrix_II.py
#!/usr/bin/python

class Solution(object):
    def generateMatrix(self, n):
        left, top = [0] * 2
        right, down = [n - 1] * 2
        direction = 0
        res = [[0] * n for i in range(n)]
        number = 1
        while 1:
        	if direction == 0:
        		for i in range(left, right + 1):
        			res[top][i] = number
        			number += 1       			
        		top += 1
        	if direction == 1:
        		for i in range(top, down + 1):
        			res[i][right] = number
        			number += 1
        		right -= 1
        	if direction == 2:
        		for i in range(right, left - 1, -1):
        			res[down][i] = number
        			number += 1
        		down -= 1
        		
        	if direction == 3:
        		for i in range(down, top - 1, -1):
        			res[i][left] = number
        			number += 1
        		left += 1
        	if number == n ** 2 + 1:
        		break
        	direction = (direction + 1) % 4
        return res
        	





def main():
    s = Solution()
    print s.generateMatrix(3)

if __name__ == "__main__":
    main()